File: python/hush-providers/hush/providers/auth/factory.py
"""Factory function for creating authentication providers."""

from .config import KeycloakTokenConfig
from .keycloak import KeycloakTokenProvider


def create_auth(config: KeycloakTokenConfig) -> KeycloakTokenProvider:
    """Create a KeycloakTokenProvider from config.

    Args:
        config: KeycloakTokenConfig instance.

    Returns:
        KeycloakTokenProvider instance.
    """
    return KeycloakTokenProvider(config)

File: rust/operonx/src/providers/auth/factory.rs
//! `create_auth` — build a [`KeycloakTokenProvider`] from its config.
//!
//! Mirrors Python [`operonx/providers/auth/factory.py`](../../../../../operonx/providers/auth/factory.py).
//! Per plan §5b.6 the factory is **Keycloak-only** — no multi-type
//! dispatcher. Expand only when a second auth provider is added.

use std::sync::Arc;

use super::config::KeycloakTokenConfig;
use super::keycloak::KeycloakTokenProvider;
use crate::core::exceptions::OperonError;

/// Construct a [`KeycloakTokenProvider`] and start its background refresh
/// loop.
pub fn create_auth(
    config: KeycloakTokenConfig,
) -> Result<Arc<KeycloakTokenProvider>, OperonError> {
    let provider = Arc::new(KeycloakTokenProvider::new(config));
    provider.start();
    Ok(provider)
}
